src/DefaultBuilder/src/WebApplicationBuilder.cs

```diff
@@ -141,6 +142,25 @@ internal WebApplicationBuilder(WebApplicationOptions options, Action<IHostBuilde
         /// <returns>A configured <see cref="WebApplication"/>.</returns>
         public WebApplication Build()
         {
+            // The `WebRoot` is the only host configuration that can be modified after
+            // the `WebApplicationBuilder` is instantiated. When the `WebRoot` is set
+            // we want to create a matching file provider that is used in DI to resolve
+            // the full paths for static assets in the application. We do this in the
+            // `Build` invocation because we want to respect only the last of the
+            // WebRootPaths configured on the application. So, when the following code
+            // is called, a directory and provider will only be created for C.
+            // builder.WebHost.UseSettings("WEBROOT", "A");
+            // builder.WebHost.UseSettings("WEBROOT", "B");
+            // builder.WebHost.UseSettings("WEBROOT", "C");
+            if (!string.IsNullOrEmpty(Environment.WebRootPath))
+            {
+                if (!Directory.Exists(Environment.WebRootPath))
+                {
+                    Directory.CreateDirectory(Environment.WebRootPath);
+                }
+                Environment.WebRootFileProvider = new PhysicalFileProvider(Environment.WebRootPath);
```

Comment:
Do we want this to be set if somebody accesses it before calling build? I dunno how meaningful it is.

Reply:
From a grep.app on this, the `WebRootFileProvider` is access in either:

1. Extension methods on the `IApplicationBuilder` (e.g. `UseBlazorFrameworkFiles`)
2. Extension methods on the `IServiceCollection` (e.g. used in some 3rd party libraries for CSS/JS bundling)

1 is fine. 2 might present some issue if the user does something like this.

```chsarp
var builder = WebApplication.CreateBuilder();
builder.Services.SomeMethodThatUsesWebRootFileProvider(); // Uses NullFileProvider
builder.WebHost.UseWebRoot("wwwwroot2");
builder.Services.SomeOtherMethodThatUsesWebRootFileProvider(); // Uses NullFileProvider
var app = builder.Build();
app.UseStaticFiles(); // Uses PhysicalFileProvider { Root = wwwroot2 }
```

It gets a little bit more tricky if there is a `wwwroot` directory already in the content root path at which point the behavior would be:

```chsarp
var builder = WebApplication.CreateBuilder();
builder.Services.SomeMethodThatUsesWebRootFileProvider(); // Uses PhysicalFileProvider { Root = wwwroot }
builder.WebHost.UseWebRoot("wwwwroot2");
builder.Services.SomeOtherMethodThatUsesWebRootFileProvider(); // Uses PhysicalFileProvider { Root = wwwroot }
var app = builder.Build();
app.UseStaticFiles(); // Uses PhysicalFileProvider { Root = wwwroot2 }
```

Possible mitigations for this include.

1. Set `WebRootFileProvider` in `UseSettings` only if the directory specified already exists. Matches the default initialization logic for the WebHostEnvironment.
2. Set `WebRootFileProvider` in `UseSettings` all the time without creating a new directory.
3. Make no change.